linux/src/completion_port_linux.c

```diff
@@ -0,0 +1,434 @@
+        result = NULL;
+    }
+all_ok:
+    return result;
+}
+
+void completion_port_inc_ref(COMPLETION_PORT_HANDLE completion_port)
+{
+    // Codes_COMPLETION_PORT_LINUX_11_006: [ If completion_port is NULL, completion_port_inc_ref shall return. ]
+    if (completion_port == NULL)
+    {
+        LogError("Invalid arguments: COMPLETION_PORT_HANDLE completion_port=%p", completion_port);
+    }
+    else
+    {
+        // Codes_COMPLETION_PORT_LINUX_11_007: [ Otherwise completion_port_inc_ref shall increment the internally maintained reference count. ]
+        INC_REF(COMPLETION_PORT, completion_port);
+    }
+}
+
+void completion_port_dec_ref(COMPLETION_PORT_HANDLE completion_port)
+{
+    // Codes_COMPLETION_PORT_LINUX_11_008: [ If completion_port is NULL, completion_port_dec_ref shall return. ]
+    if (completion_port == NULL)
+    {
+        LogError("Invalid arguments: COMPLETION_PORT_HANDLE completion_port=%p", completion_port);
+    }
+    else
+    {
+        // Codes_COMPLETION_PORT_LINUX_11_009: [ completion_port_dec_ref shall decrement the reference count for completion_port. ]
+        if (DEC_REF(COMPLETION_PORT, completion_port) == 0)
+        {
+            // Codes_COMPLETION_PORT_LINUX_11_010: [ If the reference count reaches 0, completion_port_dec_ref shall do the following: ]
+            int32_t value;
+            // Codes_COMPLETION_PORT_LINUX_11_011: [ - wait for the ongoing call count to reach zero. ]
+            while ((value = interlocked_add(&completion_port->pending_calls, 0)) != 0)
+            {
+                (void)wait_on_address(&completion_port->pending_calls, value, UINT32_MAX);
+            }
+
+            // Codes_COMPLETION_PORT_LINUX_11_012: [ - increment the flag signaling that the threads can complete. ]
+            interlocked_increment(&completion_port->worker_thread_continue);
+
+            // Codes_COMPLETION_PORT_LINUX_11_013: [ - close the epoll object. ]
+            if (close(completion_port->epoll) != 0)
+            {
+                LogErrorNo("failure closing epoll");
+            }
+
+            int dont_care;
+            // Codes_COMPLETION_PORT_LINUX_11_014: [ - close the thread by calling ThreadAPI_Join. ]
+            if (ThreadAPI_Join(completion_port->thread_handle, &dont_care) != THREADAPI_OK)
+            {
+                LogError("Failure joining thread");
+            }
+
+            // Codes_COMPLETION_PORT_LINUX_11_015: [ - then the memory associated with completion_port shall be freed. ]
+            do
+            {
+                int32_t current_val = interlocked_compare_exchange(&completion_port->recv_data_access, 1, 0);
+                if (current_val == 0)
+                {
+                    break;
+                }
+                else
+                {
+                    // Do Nothing
+                }
+                (void)wait_on_address(&completion_port->recv_data_access, current_val, UINT32_MAX);
+            } while (true);
```

Comment:
this seems to be "acquiring" the "lock" during destroy after waiting on all the threads to exit. Seems like it shouldn't be necessary (how is another thread accessing this?)